### PR TITLE
[FIX] website_sale: fix zero priced product rule

### PR DESCRIPTION
**Pull request by adrienmilis**
In website_sale, there is a setting to prevent the sale of zero priced products. It works fine but triggers also on order of free events, free event booths and free courses.

This is unwanted behavior and a check has been added to still allow those three product types to be sold for 0.

Task-3388254

### CODE REVIEW

File: addons/website_sale/models/sale_order.py
Comment (jeh-odoo) [inline addons/website_sale/models/sale_order.py:195]: I think that we could have a model method to get the list of white listed product types and update it in each module when we need to. I don't think that updating website_sale each time when we want to update that list is good. It will be a bit more verbose but at least we avoid to add stuff not related to the website_sale module. And it could perhaps be useful for custom modules.

**Review by jeh-odoo — COMMENTED** [review]
Hello @adrienmilis 

I think that we should avoid to add stuff not related to website_sale. I propose to add a method to get the list of product type for which we ignore the `prevent_zero_price_sale` and update it in the module where we need to.

Cheers!

**Comment by adrienmilis** [comment]
Hello @jeh-odoo,
I changed it as requested

File: addons/event_booth_sale/models/product.py
Comment (jeh-odoo) [inline addons/event_booth_sale/models/product.py:30]: If it's only when we have a website then it would be better in `website_event_booth_sale`

File: addons/event_sale/models/product.py
Comment (jeh-odoo) [inline addons/event_sale/models/product.py:29]: Same here would be better in `website_event_sale`

**Review by jeh-odoo — COMMENTED** [review]
Hey @adrienmilis 
Thanks for the changes!
Just two small comments 😄 
Could also be great to have a small test, to check that `_cart_update()` does raise correctly the error when it needs to and ignore it otherwise. It should not be too hard, you can get a little inspiration in `test_pricelist_different_currency()` if you want.

Cheers!

**Comment by adrienmilis** [comment]
@jeh-odoo 
Where do you want a test? In just one of the three locations where there is an override?

**Comment by adrienmilis** [comment]
Or with a mock of the function?

**Comment by jeh-odoo** [comment]
@adrienmilis Hum I would put the test in website_sale with a mock of the function for a test product type

**Comment by adrienmilis** [comment]
Hello @jeh-odoo,
I updated with the verbose version and a test
Let me know what you think :)

File: addons/website_sale/tests/test_website_sale_cart.py
Comment (jeh-odoo) [inline addons/website_sale/tests/test_website_sale_cart.py:91]:
```suggestion
        with patch.object(ProductTemplate, '_get_product_types_possible_zero_price_sale', ['service']):
```

**Review by jeh-odoo — COMMENTED** [review]
Hello @adrienmilis 
Thanks for the changes!
Last suggestions and we should be good to go :slightly_smiling_face: 

Cheers!

**Comment by adrienmilis** [comment]
Hello @jeh-odoo,
Made the latest changes and branch is green

**Review by Feyensv — APPROVED** [review]
Makes sense to fix those flows indeed :+1: .

Note that I dunno if we should bypass the check, or if we should hide/forbid the registration to those events/courses instead.

I've learned quite recently that some customers use this setting to force customers to be logged in before buying anything (with a pricelist setting all prices to 0 on the public partner), but I'm not sure that it would make sense for events/courses as well :thinking:

File: addons/website_sale/tests/test_website_sale_cart.py
Comment (adrienmilis) [reply addons/website_sale/tests/test_website_sale_cart.py:69]: Wdym more useful? It'd be redundant to test the same function in each separate module ; rather, it's better to test in the same module where the logic is defined and ensure the function works with any kind of usage

**Review by awa-odoo — COMMENTED** [review]
Hello @Feyensv 

Was going to ask specifically for you to check but you beat me to it :smile: 

To answer your review questions, having free courses / events seem like an intended behavior of those modules (maybe more arguable for booths). So I think those should be excluded from this check (instead of preventing people to buy them).

Of course this is a behavior decision, so it can be re-discussed with the POs if you don't think it's a good idea.

About customers using this feature to prevent people from buying things before logging in... this seems like a side-effect more than something intended, so not sure about that one ^^

I addressed the various comments.

Thanks for the review & cheers!

@adrienmilis I re-took the branch to make the final changes before the merge, hope you don't mind :slightly_smiling_face:

File: addons/website_sale/tests/test_website_sale_cart.py
Comment (Feyensv) [reply addons/website_sale/tests/test_website_sale_cart.py:69]: From a technical POV, it's nice to test the architecture/api, like it's done here.
But what we fix in this PR is not the behavior of the api, but specific business flows.

This test would still work even if I remove your overrides (even if it would be dumb, I agree, but let's say it's done by mistake), so we could potentially break again what you're fixing without making your tests red (unless we remove the base method itself).

Anyway, I won't block this PR for that, I wouldn't say that this test is wrong or bad, but it's not enough if you want to cover your overrides.

**Review by Feyensv — APPROVED** [review]
> To answer your review questions, having free courses / events seem like an intended behavior of those modules (maybe more arguable for booths). So I think those should be excluded from this check (instead of preventing people to buy them).
> 
> Of course this is a behavior decision, so it can be re-discussed with the POs if you don't think it's a good idea.

It's fine for me, I was just making sure that the different approaches were considered ;).

> About customers using this feature to prevent people from buying things before logging in... this seems like a side-effect more than something intended, so not sure about that one ^^

Well, I didn't hear of any such feedbacks for event/slides specifically, so I guess it shouldn't annoy more people than the ones annoyed by the current situation :D.

Furthermore, I believe it's still better to take a clear stance and handle this setting in events/slides (even if it means not considering it) instead of throwing an error to the customer that didn't do anything wrong.

**Comment by awa-odoo** [comment]
@robodoo r+

**Pull request closed without merging**
